evm-typegen/src/main.ts

```diff
@@ -0,0 +1,153 @@
+import { program } from "commander";
+import process from "process";
+import path from "path";
+import fs from "fs";
+import { Output } from "@subsquid/util-internal-code-printer";
+import { AbiCoder } from "@ethersproject/abi";
+
+export function run(): void {
+    program.description(`
+Generates TypeScript definitions for evm log events
+for use within substrate-processor mapping handlers.
+    `.trim())
+        .option('--input <path>', 'path of JSON abi file')
+        .option('--output <path>', 'path for output typescript file');
+
+    program.parse();
+
+    const options = program.opts();
+    const inputPath = options.input;
+    const outputPath = options.output;
+
+    try {
+        generateTsFromAbi(inputPath, outputPath);
+    } catch(e: any) {
+        printError(e);
+        process.exit(1);
+    }
+}
+
+function printError(err: any) {
+    console.error(`substrate-typegen ${err.toString()}`);
+}
+
+function generateTsFromAbi(inputPathRaw: string, outputPathRaw: string): void {
+    const inputPath = path.parse(inputPathRaw);
+    const outputPath = path.parse(outputPathRaw);
+
+    if (inputPath.ext !== ".json") {
+        throw new Error("invalid input file extension");
+    }
+
+    if (outputPath.ext !== ".ts") {
+        throw new Error("invalid output file extension");
+    }
+
+    const rawABI = JSON.parse(fs.readFileSync(inputPathRaw, { encoding: "utf-8" }));    
+
+    const output = new Output();
+
+    output.line("import { Interface } from \"@ethersproject/abi\";");
+    output.line("import ethers from \"ethers\";");
+    output.line("import { EvmLogHandlerContext } from \"@subsquid/substrate-evm-processor\";");
+    output.line(`import inputJson from "${path.resolve(inputPathRaw)}";`);
+    output.line("");
+    output.line("const abi = new Interface(inputJson);");
+    output.line("");
+
+    const eventTypeIndexes: { [key: string]: number } = {};
+
+    const abiEvents = rawABI.filter((decl: any): boolean => decl.type === "event").map((decl: any): any => {
+        let signature = `${decl.name}(`;
+    
+        if(decl.inputs.length > 0) {
+            signature += decl.inputs[0].type;
+        }
+    
+        for (let i=1; i<decl.inputs.length; ++i) {
+            const input = decl.inputs[i];
+            signature += `,${input.type}`;
+        }
+    
+        signature += ")";
+    
+        decl.signature = signature;
+        if (eventTypeIndexes[decl.name] === undefined) {
+            eventTypeIndexes[decl.name] = 0;
+        } else {
+            eventTypeIndexes[decl.name]++;
+        }
+        decl.eventTypeName = decl.name + eventTypeIndexes[decl.name].toString();
```

Comment:
This will work regardless because it is using the same order to generate the typenames and the decode implementations. User might get different typenames when he regenerates the code with a different abi file but this will be very unlikely, also user probably won't be typing the output explicitly unless he is passing it to other functions

Reply:
I guess doing the long signature based one is ok as well but it is harder to type in case they do want to pass it around